### ISSUE TITLE
Internal Drag and Drop of any Inline Wrapped Content Causes Content to Disappear in CKEDITOR4

### ISSUE THREAD

**Issue by brianikim** [ckeditor/ckeditor-dev#2595]
## Type of report

Bug

## Provide detailed reproduction steps (if any)

1. Go to nightly.ckeditor.com/full/samples/
2. Use Backspace to clear the editor completely so that paragraph formatting returns to Normal as well.
3. Click on Image in the toolbar.
4. In Image Info, provide the address for an image in the URL section. Please use a small image so that the drag and drop can easily be achieved.
5. Click on Link and type in a link to wrap the image with. I just used www.google.com
6. Click OK
7. Resize the Editor only to the point that the scroll bar disappears.
8. Do an internal drag and drop of the image. (You might have to do this multiple times depending on when the error condition is met for it to be triggered).

See following for a demonstration of the reproduction steps (Open GIF in new tab to see better):

![dragdrop](https://user-images.githubusercontent.com/35237986/48236691-4d769d80-e378-11e8-824b-73ead49a61ce.gif)

### Expected result

If there are no other content inside the editor other than the image, I expect the internal drag&drop to behave in such a way that the image stays in its original location. This behavior can be observed in the internal drag&drop of just an image.

### Actual result

The image disappears and the following errors are thrown in the console log:

> Uncaught TypeError: Cannot read property 'type' of null
>     at window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.dom.range.setStart (ckeditor.js:165)
>     at window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.dom.range.setStartAfter (ckeditor.js:166)
>     at window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.dom.range.setStartAt (ckeditor.js:167)
>     at window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.dom.range.moveToPosition (ckeditor.js:164)
>     at window.CKEDITOR.window.CKEDITOR.dom.CKEDITOR.dom.range.splitElement (ckeditor.js:170)
>     at ckeditor.js:386
>     at $.insertHtml (ckeditor.js:354)
>     at $.<anonymous> (ckeditor.js:364)
>     at a.d (ckeditor.js:10)
>     at a.<anonymous> (ckeditor.js:11)

**Note** that this issue can be reproducible even with other content inside the editor. The issue is simply easier to reproduce in the editor if it only contains the image wrapped in an anchor tag.

## Other details

* Browser: Best reproduced on Chrome. Mileage may vary on Firefox. I have not tested on IE.
* OS: Ubuntu 18.04. Issue should be reproducible on Windows as well.
* CKEditor version: The latest version that is used in the sample nightly build. (Refer to step 1 for link)
* Installed CKEditor plugins: whatever plugins that are included with the sample nightly build. (Refer to step 1 for link)

**Comment by blzaugg** [comment]
From About CKEditor 4:
> CKEditor 4.11.0 (Full) (revision 65f3768)

**Comment by engineering-this** [comment]
I can reproduce it with described test steps.

**Comment by brianikim** [comment]
Hello @engineering-this ,

Any updates for this issue? The current behavior pretty much prevents any effective use of wrappings images with links.

**Comment by blzaugg** [comment]
**Simplified replication steps:**
1. Select any inline (`<a>`, `<strong>`, etc.) wrapped content.
2. Drag selection to the boundaries of the editor. Works best using the nearest border.
3. Notice mouse cursor shows the selection-drag is valid (versus an invalid cursor).
4. Release mouse.

**Comment by blzaugg** [comment]
![wrapped content selection drag 1](https://user-images.githubusercontent.com/1626739/48441746-27b71300-e741-11e8-88e1-101d55a89d75.gif)

**Comment by blzaugg** [comment]
I've been looking into this some. I've traced the issue to `node = path.contains( isInline, false, 1 )`

It seems to want to split inline text elements, based on where you start the drag.

**Comment by engineering-this** [comment]
Thanks for sharing your findings. This will be helpful when fixing this issue. Although we don't have any updates on this yet.

**Comment by blzaugg** [comment]
Also, `node`s `parentNode` ends up being a `NODE_DOCUMENT_FRAGMENT`. I would have expected it's parent to be a `NODE_ELEMENT`.

**Comment by blzaugg** [comment]
parentNode: #document-fragment
```
<a data-cke-saved-href=​"https:​/​/​ckeditor.com" href=​"https:​/​/​ckeditor.com">​</a>​
<a data-cke-saved-href=​"https:​/​/​ckeditor.com" href=​"https:​/​/​ckeditor.com">​"CK""Editor"</a>
```

**Comment by blzaugg** [comment]
Possible solution:

https://github.com/ckeditor/ckeditor-dev/blob/major/plugins/clipboard/plugin.js#L1927
```diff

			editor.fire( 'saveSnapshot' );
			editor.fire( 'lockSnapshot', { dontUpdate: 1 } );
+			
+			// Include inline element if possible.
+			dragRange.enlarge( CKEDITOR.ENLARGE_INLINE, 1 );

			if ( CKEDITOR.env.ie && CKEDITOR.env.version < 10 ) {
				this.fixSplitNodesAfterDrop(
					dragRange,
```

It looks like in the middle of the drop logic, the inline element gets inconsistently handled by range and drop target. So the drop target bookmark ends up being inside the drag range bookmarks rather than in front of the range. Then the drop target bookmark gets removed during cleanup, causing the issue.

The issue is visible if you debug/break at: 
https://github.com/ckeditor/ckeditor-dev/blob/major/core/editable.js#L727

Inspect the DOM (some debug attributes added for clarity):
```html
<p>
	I'm an instance of 
	<span data-cke-bookmark="1" DEBUG_startNode DEBUG_range_bookmark" style="display: none;">&nbsp;</span>
	<a data-cke-saved-href="https://ckeditor.com" href="https://ckeditor.com">
		<span data-cke-bookmark="1" DEBUG_startNode DEBUG_drop_target_bookmark style="display: none;">&nbsp;</span>
		CKEditor
	</a>
	<span data-cke-bookmark="1" DEBUG_endNode" DEBUG_range_bookmark" style="display: none;">&nbsp;</span>
	.
</p>
```

**Comment by shortPause** [comment]
Hi CKEditor Team, I'm facing the similar issue under version 4.9 of the CKEditor. The drag and drop operation inside the content box is getting disappeared on Edge browser i.e. as soon as I drag a text from one position to another and try to drop, it gets disappeared. The Edge browser version is as below:

Microsoft Edge 42.17134.1.0
Microsoft EdgeHTML 17.17134

Please advise for the fix or the workaround!

**Comment by shortPause** [comment]
Hi Team, any update would be much appreciated.

**Comment by Comandeer** [comment]
It seems that the issue was fixed by #2803, therefore the fix will be officially released in 4.11.3.
